server/src/main/java/org/uiautomation/ios/wkrdp/message/MessageFactory.java

```diff
@@ -14,61 +14,186 @@
 
 package org.uiautomation.ios.wkrdp.message;
 
-import org.apache.commons.io.IOUtils;
-import org.dom4j.Document;
-import org.dom4j.DocumentException;
-import org.dom4j.Node;
-import org.dom4j.io.SAXReader;
-
 import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.util.Arrays;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
 import java.util.logging.Level;
 import java.util.logging.Logger;
 
+import org.apache.commons.io.IOUtils;
+import org.dom4j.Document;
+import org.dom4j.DocumentException;
+import org.dom4j.Node;
+import org.dom4j.io.SAXReader;
+
 /**
- * Created with IntelliJ IDEA. User: freynaud Date: 17/01/2013 Time: 15:08 To change this template
- * use File | Settings | File Templates.
+ * Created with IntelliJ IDEA. User: freynaud Date: 17/01/2013 Time: 15:08 To change this template use File | Settings |
+ * File Templates.
  */
 public class MessageFactory {
 
-  private static final Logger log = Logger.getLogger(MessageFactory.class.getName());
-  private final
-  Map<String, Class<? extends BaseIOSWebKitMessage>>
-      types =
-      new HashMap<String, Class<? extends BaseIOSWebKitMessage>>();
+  private static final Logger LOG = Logger.getLogger(MessageFactory.class.getName());
+
+  private static final List<String> DEFAULT = Arrays.asList("Default");
+
+  private static final List<String> IOS7_TYPES = Arrays.asList("7.0.3", "7.1");
+
+  private static final List<String> IOS8_TYPES = Arrays.asList("8.0", "8.1");
+
+  private final Map<List<String>, Map<String, Class<? extends BaseIOSWebKitMessage>>> iOSTypesMap;
+
+  private String iOSVersion;
+
+  private ReentrantLock versionDeterminingLock;
+
+  private Condition determiningVersion;
 
   public MessageFactory() {
-    types.put("_rpc_reportSetup:", ReportSetupMessage.class);
-    types.put("_rpc_reportConnectedApplicationList:", ReportConnectedApplicationsMessage.class);
-    types.put("_rpc_applicationSentListing:", ApplicationSentListingMessage.class);
-    types.put("_rpc_applicationSentData:", ApplicationDataMessage.class);
-    types.put("_rpc_applicationConnected:", ApplicationConnectedMessage.class);
-    types.put("_rpc_applicationDisconnected:", ApplicationDisconnectedMessage.class);
+    iOSTypesMap = new HashMap<>();
+    versionDeterminingLock = new ReentrantLock();
+    determiningVersion = versionDeterminingLock.newCondition();
+    populateDefaultTypes();
+    populateIOS7Types();
+    populateIOS8Types();
+  }
+
+  private void populateDefaultTypes() {
+    Map<String, Class<? extends BaseIOSWebKitMessage>> types = new HashMap<>();
+    types.put("_rpc_reportSetup:", org.uiautomation.ios.wkrdp.message.ReportSetupMessageImpl.class);
```

Comment:
Any specific reason behind why we are using a fully qualified package name here instead of resorting to adding an import statement and then just using the class's simple name.